Yuanqing/feature records list view

Yqzzxlj opened the pull request:
List-view page for training-records.

Yqzzxlj committed:
Add the list-view page of training-records#31
update the interfaces for record and event

Yqzzxlj committed:
fix the code style#31

Yqzzxlj committed:
fixed the style of code#31

Yqzzxlj committed:
update the testing code#31

Time1ess commented on src/app/components/sidebar/sidebar.component.ts:
```diff
@@ -16,6 +16,7 @@ export const ADMIN_ROUTE_ITEMS: RouteInfo[] = [
 export const REGULAR_USER_ROUTE_ITEMS: RouteInfo[] = [
     { path: '/dashboard', title: '首页', icon: 'dashboard', class: '' },
     { path: '/training-record/entry', title: '培训记录填报', icon: 'create', class: '' },
+    { path: '/training-record/list-view', title: '培训记录列表', icon: 'list', class: '' },
```
`/training-record/records` is preferred, since it's consistent with backend server.

Time1ess commented on src/app/interfaces/event.ts:
```diff
@@ -24,6 +24,8 @@ export interface CampusEventRequest extends CampusEvent {
 /** RESTful response interface for CampusEvent. */
 export interface CampusEventResponse extends CampusEvent {
   id: number;
+  create_time: string;
```
Since `create_time` and `update_time` exist in both `CampusEvent` and `OffCampusEvent`, you can move these fields to `Event`, and mark them as optional, for example:
```
interface Event {
  name: string;
  time: string;
  location: string;
  create_time?: string;
  update_time?: string;
  num_hours: number;
  num_participants: number;
}
```

Time1ess commented on src/app/modules/training-record/components/list-view/list-view.component.html:
```diff
@@ -0,0 +1,57 @@
+                    <td class="col-8 col-sm-2">{{ record.off_campus_event.location }}</td>
+                    <td class="col-8 col-sm-2">{{ record.off_campus_event.num_hours }}</td>
+                    <td class="col-8 col-sm-2">{{ record.off_campus_event.num_participants }}</td>
+                    <td *ngIf="record.status === 0" class="col-8 col-sm-2 text-justify">未提交</td>
```
It's not clean to display human readable status in this way, please use `Pipe` to fulfill the need. See https://angular.cn/guide/pipes

Time1ess commented on src/app/modules/training-record/components/list-view/list-view.component.html:
```diff
@@ -0,0 +1,57 @@
+                    <td class="col-8 col-sm-2">{{ record.campus_event.location }}</td>
+                    <td class="col-8 col-sm-2">{{ record.campus_event.num_hours }}</td>
+                    <td class="col-8 col-sm-2">{{ record.campus_event.num_participants }}</td>
+                    <td *ngIf="record.status === 0" class="col-8 col-sm-2 text-justify">未提交</td>
```
Ditto.

Time1ess commented on src/app/modules/training-record/components/list-view/list-view.component.spec.ts:
```diff
@@ -0,0 +1,112 @@
+        {
+          provide: RecordService,
+          useValue: {
+            getRecords,
```
You can remove the jasmine.Spy variable `getRecords`, instead do like this:

```
useValue: {
  getRecords: () => getRecord$,
}
```

Time1ess commented on src/app/modules/training-record/components/list-view/list-view.component.ts:
```diff
@@ -0,0 +1,65 @@
+  templateUrl: './list-view.component.html',
+  styleUrls: ['./list-view.component.css']
+})
+export class ListViewComponent implements OnInit {
```
Rename `ListViewComponent` since it's a general name, consider using `RecordListComponent`, and replace all related names.

Time1ess commented on src/app/modules/training-record/components/list-view/list-view.component.ts:
```diff
@@ -0,0 +1,65 @@
+    /** The data to be displayed */
+    records: RecordResponse[] = [];
+    /** How to sort the columns */
+    displayedColumns: string[] = ['program_name', 'time', 'location', 'num_hours', 'num_participants', 'status'];
```
You're not using MatTable, so this variable is redundant.

Time1ess commented on src/app/modules/training-record/components/list-view/list-view.component.ts:
```diff
@@ -0,0 +1,65 @@
+    /** Indicate data loading status */
+    isLoadingResults = true;
+
+    pageSize = environment.PAGINATION_SIZE;
```
Instead of defining `pageSize`, use `environment.PAGINATION_SIZE` directly.

Time1ess commented on src/app/modules/training-record/services/record.service.spec.ts:
```diff
@@ -79,4 +89,31 @@ describe('RecordService', () => {
+    req.flush({count: 2, results: [{}, {}]});
+  });
+
+  it('should ignore errors during getting records', fakeAsync(() => {
```
What are you exactly testing in this unit test case? I don't see any `expect` statements. And why are you using `fakeAsync`?

Time1ess commented on src/app/modules/training-record/services/record.service.ts:
```diff
@@ -1,8 +1,9 @@
 import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
-import { Observable } from 'rxjs';
+import { Observable} from 'rxjs';
```
Asymmetric spaces.

Time1ess commented on src/app/modules/training-record/services/record.service.ts:
```diff
@@ -30,4 +31,16 @@ export class RecordService {
     return this.http.delete(`${environment.API_URL}/records/${recordID}/`);
   }
+
+  /** Get Records. */
```
Don't comment with no meaning, you are just repeating the function name.

Time1ess commented on src/app/modules/training-record/services/record.service.ts:
```diff
@@ -30,4 +31,16 @@ export class RecordService {
+
+  /** Get Records. */
+  getRecords(offset?: number, limit?: number) {
+    if (offset === undefined) offset = 0;
```
Replace with `offset = offset ||  0;`

Time1ess commented on src/app/modules/training-record/training-record-routing.module.ts:
```diff
@@ -29,6 +30,16 @@ const routes: Routes = [
         ]
+      },
+      {
+        path: 'list-view',
```
Ditto.

Time1ess reviewed (changes requested):
Please attach a snapshot of your work.

Time1ess commented on src/app/pipes/record-status.pipe.ts:
```diff
@@ -0,0 +1,11 @@
+
+@Pipe({ name: 'recordStatusDisplay' })
+export class RecordStatusDisplayPipe implements PipeTransform {
+    transform(value: number) {
```
1. I believe `value: RecordStatus` is better than `valud: number`, you can check the `src/app/enums/record-status.enum.ts`, BTW, please update the field in `Record`.
2. Use `switch` instead of multiple `if`.
3. Add a fallback solution when we encountered unexpected value, such as return '未知状态' if no match for all cases.

Time1ess commented on src/app/pipes/record-status.pipe.ts:
```diff
@@ -0,0 +1,11 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({ name: 'recordStatusDisplay' })
+export class RecordStatusDisplayPipe implements PipeTransform {
```
Add unit tests for the pipe.

Time1ess reviewed (changes requested):
The code coverage is decreased, pls fix.

Time1ess commented on src/app/pipes/record-status.pipe.ts:
```diff
@@ -1,11 +1,25 @@
-        if (value === 1) return '已提交';
-        if (value === 2) return '院系管理员已审核';
-        if (value === 3) return '学校管理员已审核';
+    transform(value: RecordStatus) {
```
Hmm, it's better, but I think we can use a map to provide <status, human_readable_status> pairs, then we don't need the `switch` any more.

Yqzzxlj commented:
1. Failed to use ENUM as the key of map, so keeped the pipe as original.
2. Force-pushed the new branch to eliminate the merging with dev-branch.

And here is the screenshot of the page.
![image](https://user-images.githubusercontent.com/43257534/53810385-c508b200-3f91-11e9-9e0a-575a5cdcb5ee.png)

Yqzzxlj committed:
fixed the code as recommended
add the pipe for record-stauts#31

Time1ess commented:
I thinks this works:
```
enum Category {
    CATE_A = 0,
    CATE_B = 1,
    CATE_C = 2,
};

const m = new Map<Category, string>(
    [
        [Category.CATE_A, 'hello a'],
        [Category.CATE_B, 'hello b'],
        [Category.CATE_C, 'hello c'],
    ],
);
```

Yqzzxlj commented:
1. Corrected the column-width in html file.
2. Changed SWITCH to MAP for improving the status-pipe.

And here is the screenshot of the page.

![image](https://user-images.githubusercontent.com/43257534/53877566-7700b680-4044-11e9-89e1-abaf01c509b3.png)

Time1ess commented:
Remove `text-justify` in class attribute of `td` to improve UI.

Yqzzxlj commented:
Removed text-justify.
And here is the screenshot of the page.
![image](https://user-images.githubusercontent.com/43257534/53882185-9271be80-4050-11e9-91d3-e7b85fb91929.png)

Yqzzxlj committed:
Corrected the column-width in html file.
Changed SWITCH to MAP for improving the status-pipe.#31

Yqzzxlj committed:
Removed text-justify in class attribute of td to improve UI#31
